module/zfs/dsl_prop.c

```diff
@@ -1123,10 +1125,11 @@ dsl_prop_get_all_ds(dsl_dataset_t *ds, nvlist_t **nvp,
 		dsl_dir_name(dd, setpoint);
 		err = dsl_prop_get_all_impl(mos,
 		    dsl_dir_phys(dd)->dd_props_zapobj, setpoint, flags, *nvp);
-		if (err)
-			break;
+		if (err) {
+			nvlist_free(*nvp);
+			return (err);
+		}
 	}
-out:
```

Comment:
The function is already structured such that everything returns through the `out:` tag so how about something like this instead.  This has a few advantages.:

* Minimal modification to the existing `dsl_prop_get_all_ds()` implementation.  The existing flow is preserved which minimizes the change on merge conflicts or until this fix is  merged in to OpenZFS.

* Easier to read.  It's clear that when exiting if there was an error the *nvp must be freed.

* The *nvp should be set to NULL on error and not a recently freed address.  The `zfs_ioc_recv_impl()` function looks like it depends on this behavior.

```diff
diff --git a/module/zfs/dsl_prop.c b/module/zfs/dsl_prop.c
index 2483600..294bd30 100644
--- a/module/zfs/dsl_prop.c
+++ b/module/zfs/dsl_prop.c
@@ -1127,6 +1127,11 @@ dsl_prop_get_all_ds(dsl_dataset_t *ds, nvlist_t **nvp,
                        break;
        }
 out:
+       if (err) {
+               nvlist_free(*nvp);
+               *nvp = NULL;
+       }
+
        return (err);
 }
 
```